In [31]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from numpy._core.defchararray import upper

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [32]:
Direccion_guardado = "/workspaces/Marcos-J-Gomes_https-github.com-4GeeksAcademy-machine-learning-python_LinealRegression/data/processed"


X_train_sel_CON_outliers= pd.read_csv(f"{Direccion_guardado}/X_train_sel_CON_outliers_clean.csv")
X_test_sel_CON_outliers= pd.read_csv(f"{Direccion_guardado}/X_test_sel_CON_outliers_clean.csv")
X_train_sel_CON_outliers_norm=pd.read_csv(f"{Direccion_guardado}/X_train_sel_CON_outliers_clean.csv")
X_test_sel_CON_outliers_norm=pd.read_csv(f"{Direccion_guardado}/X_test_sel_CON_outliers_clean.csv")
X_train_sel_CON_outliers_scal=pd.read_csv(f"{Direccion_guardado}/X_train_sel_CON_outliers_scal_clean.csv")
X_test_sel_CON_outliers_scal=pd.read_csv(f"{Direccion_guardado}/X_test_sel_CON_outliers_scal_clean.csv")
X_train_sel_SIN_outliers=pd.read_csv(f"{Direccion_guardado}/X_train_sel_SIN_outliers_clean.csv")
X_test_sel_SIN_outliers=pd.read_csv(f"{Direccion_guardado}/X_test_sel_SIN_outliers_clean.csv")
X_train_sel_SIN_outliers_norm=pd.read_csv(f"{Direccion_guardado}/X_train_sel_SIN_outliers_norm_clean.csv")
X_test_sel_SIN_outliers_norm=pd.read_csv(f"{Direccion_guardado}/X_test_sel_SIN_outliers_norm_clean.csv")
X_train_sel_SIN_outliers_scal=pd.read_csv(f"{Direccion_guardado}/X_train_sel_SIN_outliers_scal_clean.csv")
X_test_sel_SIN_outliers_scal=pd.read_csv(f"{Direccion_guardado}/X_test_sel_SIN_outliers_scal_clean.csv")
y_train = pd.read_excel(f"{Direccion_guardado}/y_train.xlsx")
y_test  = pd.read_excel(f"{Direccion_guardado}/y_test.xlsx")

target= "charges"

# Regresion Lineal

normal

In [33]:
datasets = {
    "CON_outliers": (X_train_sel_CON_outliers.drop(target, axis = 1), X_test_sel_CON_outliers.drop(target, axis = 1)),
    "SIN_outliers": (X_train_sel_SIN_outliers.drop(target, axis = 1), X_test_sel_SIN_outliers.drop(target, axis = 1))
}

Resultados = {}

for dataset_, (X_train, X_test) in datasets.items():
    model = LinearRegression()
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados[dataset_] = {
        "Model" : model,
        "R2": r2_score(y_test, y_pred),
        "MAE": mean_absolute_error(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred)),
        
    }

mejor_dataset = max(Resultados, key=lambda x: Resultados[x]["R2"])
mejor_r2 = Resultados[mejor_dataset]["R2"]
mejor_modelo = Resultados[mejor_dataset]["Model"]

print("Resultados por dataset:")
for k, v in Resultados.items():
    print(f"{k:25s} -> R2: {v['R2']:.5f} | MAE: {v['MAE']:.4f} | RMSE: {v['RMSE']:.4f}")

print("\nMejor dataset:", mejor_dataset)
print("Mejor R2:", mejor_r2)
print("\nParámetros del mejor modelo:")
print("Intercepto (a):", mejor_modelo.intercept_)
print("Coeficientes (b):",mejor_modelo.coef_)


Resultados por dataset:
CON_outliers              -> R2: 0.80455 | MAE: 4198.5926 | RMSE: 5992.8751
SIN_outliers              -> R2: 0.80468 | MAE: 4203.1624 | RMSE: 5990.8592

Mejor dataset: SIN_outliers
Mejor R2: 0.8046845735160492

Parámetros del mejor modelo:
Intercepto (a): [11716.58588215]
Coeficientes (b): [[   248.70418631    308.4713513     536.4255666  -23046.43330225]]


In [34]:
X_train_sel_CON_outliers

age       bmi  children  smoker_n      charges
0    -1.157680 -0.996928 -0.907908  0.500292   2396.09590
1    -1.300619 -0.792762  0.766904  0.500292   3279.86855
2     0.914926  1.154664  0.766904  0.500292  33471.97189
3     1.701087  1.806837 -0.907908  0.500292  13405.39030
4     0.557580 -0.651417  0.766904  0.500292   9715.84100
...        ...       ...       ...       ...          ...
1064  0.843457  0.730628  0.766904 -1.998831  44641.19740
1065 -0.871804  2.539188  0.766904  0.500292   3693.42800
1066 -1.372088 -1.411872 -0.070502  0.500292   1964.78000
1067 -0.085643 -0.419974  1.604310  0.500292   7151.09200
1068 -0.300050  0.879412  0.766904  0.500292   5836.52040

[1069 rows x 5 columns]

Ridge

In [35]:
datasets_ = {
    "CON_outliers": (X_train_sel_CON_outliers.drop(target, axis = 1), X_test_sel_CON_outliers.drop(target, axis = 1)),
    "CON_outliers_norm": (X_train_sel_CON_outliers_norm.drop(target, axis = 1), X_test_sel_CON_outliers_norm.drop(target, axis = 1)),
    "CON_outliers_scal": (X_train_sel_CON_outliers_scal.drop(target, axis = 1), X_test_sel_CON_outliers_scal.drop(target, axis = 1)),

    "SIN_outliers": (X_train_sel_SIN_outliers.drop(target, axis = 1), X_test_sel_SIN_outliers.drop(target, axis = 1)),
    "SIN_outliers_norm": (X_train_sel_SIN_outliers_norm.drop(target, axis = 1), X_test_sel_SIN_outliers_norm.drop(target, axis = 1)),
    "SIN_outliers_scal": (X_train_sel_SIN_outliers_scal.drop(target, axis = 1), X_test_sel_SIN_outliers_scal.drop(target, axis = 1))
}

Resultados_R = {}

for datasett, (X_train, X_test) in datasets_.items():
    model = Ridge(alpha = 0.1, max_iter = 300)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados_R[datasett] = {
        "Model" : model,
        "R2": r2_score(y_test, y_pred),
        "MAE": mean_absolute_error(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred))
    }

mejor_dataset_R = max(Resultados_R, key=lambda x: Resultados_R[x]["R2"])
mejor_r2_R = Resultados_R[mejor_dataset_R]["R2"]
mejor_modelo_R = Resultados_R[mejor_dataset_R]["Model"]

print("Resultados por dataset:")
for k, v in Resultados_R.items():
    print(f"{k:25s} -> R2: {v['R2']:.4f} | MAE: {v['MAE']:.4f} | RMSE: {v['RMSE']:.4f}")

print("\nMejor dataset:", mejor_dataset_R)
print("Mejor R2:", mejor_r2_R)

print("\nParámetros del mejor modelo:")
print("Intercepto (a):", mejor_modelo_R.intercept_)
print("Coeficientes (b):",mejor_modelo_R.coef_)

Resultados por dataset:
CON_outliers              -> R2: 0.8045 | MAE: 4198.8502 | RMSE: 5993.1495
CON_outliers_norm         -> R2: 0.8045 | MAE: 4198.8502 | RMSE: 5993.1495
CON_outliers_scal         -> R2: 0.8044 | MAE: 4199.0750 | RMSE: 5995.1941
SIN_outliers              -> R2: 0.8046 | MAE: 4204.8504 | RMSE: 5992.2870
SIN_outliers_norm         -> R2: 0.8047 | MAE: 4203.4196 | RMSE: 5991.1333
SIN_outliers_scal         -> R2: 0.8045 | MAE: 4204.0275 | RMSE: 5993.0063

Mejor dataset: SIN_outliers_norm
Mejor R2: 0.8046667007183342

Parámetros del mejor modelo:
Intercepto (a): [13030.20336929]
Coeficientes (b): [ 3479.54052937  1850.7413865    640.55542068 -9220.92703326]


Lasso

In [36]:
datasets_ = {
    "CON_outliers": (X_train_sel_CON_outliers.drop(target, axis = 1), X_test_sel_CON_outliers.drop(target, axis = 1)),
    "CON_outliers_norm": (X_train_sel_CON_outliers_norm.drop(target, axis = 1), X_test_sel_CON_outliers_norm.drop(target, axis = 1)),
    "CON_outliers_scal": (X_train_sel_CON_outliers_scal.drop(target, axis = 1), X_test_sel_CON_outliers_scal.drop(target, axis = 1)),

    "SIN_outliers": (X_train_sel_SIN_outliers.drop(target, axis = 1), X_test_sel_SIN_outliers.drop(target, axis = 1)),
    "SIN_outliers_norm": (X_train_sel_SIN_outliers_norm.drop(target, axis = 1), X_test_sel_SIN_outliers_norm.drop(target, axis = 1)),
    "SIN_outliers_scal": (X_train_sel_SIN_outliers_scal.drop(target, axis = 1), X_test_sel_SIN_outliers_scal.drop(target, axis = 1))
}

Resultados_L = {}

for datasett, (X_train, X_test) in datasets_.items():
    model = Lasso(alpha = 0.1, max_iter = 300)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    Resultados_L[datasett] = {
        "Model" : model,
        "R2": r2_score(y_test, y_pred),
        "MAE": mean_absolute_error(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred))
    }

mejor_dataset_L = max(Resultados_L, key=lambda x: Resultados_L[x]["R2"])
mejor_r2_L = Resultados_L[mejor_dataset_L]["R2"]
mejor_modelo_L = Resultados_L[mejor_dataset_L]["Model"]

print("Resultados por dataset:")
for k, v in Resultados_L.items():
    print(f"{k:25s} -> R2: {v['R2']:.4f} | MAE: {v['MAE']:.4f} | RMSE: {v['RMSE']:.4f}")

print("\nMejor dataset:", mejor_dataset_L)
print("Mejor R2:", mejor_r2_L)

print("\nParámetros del mejor modelo:")
print("Intercepto (a):", mejor_modelo_L.intercept_)
print("Coeficientes (b):",mejor_modelo_L.coef_)

Resultados por dataset:
CON_outliers              -> R2: 0.8046 | MAE: 4198.5995 | RMSE: 5992.9136
CON_outliers_norm         -> R2: 0.8046 | MAE: 4198.5995 | RMSE: 5992.9136
CON_outliers_scal         -> R2: 0.8045 | MAE: 4198.5183 | RMSE: 5993.0019
SIN_outliers              -> R2: 0.8047 | MAE: 4203.2276 | RMSE: 5990.9212
SIN_outliers_norm         -> R2: 0.8047 | MAE: 4203.1690 | RMSE: 5990.8975
SIN_outliers_scal         -> R2: 0.8047 | MAE: 4203.1073 | RMSE: 5990.9768

Mejor dataset: SIN_outliers_norm
Mejor R2: 0.804682077477434

Parámetros del mejor modelo:
Intercepto (a): [13030.20336929]
Coeficientes (b): [ 3479.79915436  1850.79901581   640.48648395 -9221.70212813]


Notamos que los modelos de Ridge y Lasso no mejoran practicamente nada al modelo Lineal base, por lo que nos quedariamos con este para trabajar:

- La edad, el BMI y el número de hijos aumentan progresivamente el coste médico

- El factor más determinante es el fumar

- No ser fumador reduce significativamente el coste anual (-23046.43)

- El modelo es coherente con la lógica médica

- El modelo explica aproximadamente el 80% de la variabilidad del coste médico anual (R2)

- Con un error medio cercano a 4.200€ (MAE)